lib/active_hash/relation.rb

```diff
@@ -7,24 +7,80 @@ class Relation
     delegate :empty?, :length, :first, :second, :third, :last, to: :records
     delegate :sample, to: :records
 
-    def initialize(klass, all_records, query_hash = nil)
+    attr_reader :conditions, :order_values, :klass, :all_records
+
+    def initialize(klass, all_records, conditions = nil, order_values = nil)
       self.klass = klass
       self.all_records = all_records
-      self.query_hash = query_hash
-      self.records_dirty = false
+      self.conditions = Conditions.wrap(conditions || [])
+      self.order_values = order_values || []
+    end
+
+    def where(conditions_hash = :chain)
+      return WhereChain.new(self) if conditions_hash == :chain
+
+      spawn.where!(conditions_hash)
+    end
+
+    class WhereChain
+      attr_reader :relation
+
+      def initialize(relation)
+        @relation = relation
+      end
+
+      def not(conditions_hash)
+        relation.conditions << Condition.new(conditions_hash).invert!
+        relation
+      end
+    end
+
+    def order(*options)
+      spawn.order!(*options)
+    end
+
+    def reorder(*options)
+      spawn.reorder!(*options)
+    end
+
+    def where!(conditions_hash, inverted = false)
+      self.conditions << Condition.new(conditions_hash)
       self
     end
 
-    def where(query_hash = :chain)
-      return ActiveHash::Base::WhereChain.new(self) if query_hash == :chain
+    def spawn
+      self.class.new(klass, all_records, conditions, order_values)
+    end
+
+    def order!(*options)
+      check_if_method_has_arguments!(:order, options)
+      self.order_values += preprocess_order_args(options)
+      self
+    end
+
+    def reorder!(*options)
+      check_if_method_has_arguments!(:order, options)
+
+      self.order_values = preprocess_order_args(options)
+      @records = apply_order_values(records, order_values)
 
-      self.records_dirty = true unless query_hash.nil? || query_hash.keys.empty?
-      self.query_hash.merge!(query_hash || {})
+      self
+    end
+
+    def records
+      @records ||= begin
+        filtered_records = apply_conditions(all_records, conditions)
+        ordered_records = apply_order_values(filtered_records, order_values) # rubocop:disable Lint/UselessAssignment
+      end
+    end
+
+    def reload
+      @records = nil # Reset records
       self
     end
 
     def all(options = {})
-      if options.has_key?(:conditions)
+      if options.key?(:conditions)
```

Comment:
You did not write this code, so probably ignore this but...

is there a reason to have the conditional here instead of an `||`?

```ruby
where(options[:conditions] || {})
```

Reply:
There would be a difference when passing an explicit false-y (`nil`, `false`, ..) value as the `:condition`, right?

But not sure what effect it has for the finders though.